scimma_admin/hopskotch_auth/models.py

```diff
@@ -202,3 +207,312 @@ def rand_username(owner: User) -> str:
+    )
+    topic = models.ForeignKey(
+        KafkaTopic,
+        on_delete=models.CASCADE,
+    )
+    operation = enum.EnumField(
+        KafkaOperation,
+        editable=False,
+    )
+
+
+# This needs to be a character which is valid neither in Kafka topic names nor in enum names
+cred_perm_encoding_separator=':'
+
+
+def encode_cred_permission(parent_id, topic_id, operation) -> str:
+    return str(parent_id)+cred_perm_encoding_separator+str(topic_id)+cred_perm_encoding_separator+str(operation)
+
+
+# returns tuples of (parent ID, topic ID, operation type)
+def decode_cred_permission(encoded: str):
+    pattern = re.compile("([^"+cred_perm_encoding_separator+"]+)"+cred_perm_encoding_separator+"([^"+cred_perm_encoding_separator+"]+)"+cred_perm_encoding_separator+"([^"+cred_perm_encoding_separator+"]+)")
+    match = re.fullmatch(pattern, encoded)
+    if match is None or len(match.groups())!=3:
+        raise ValueError("Invalid encoded credential permission")
+    try:
+        op = KafkaOperation[match[3]]
+    except KeyError:
+        raise ValueError("Invalid encoded credential permission")
+    return (match[1],match[2],op)
+
+
+# determine whether credential permission can be validly derived from the given group permission
+def supportingPermission(group_perm,cred_perm):
+    if cred_perm.topic != group_perm.topic:
+        return False
+    return group_perm.operation == KafkaOperation.All \
+      or cred_perm.operation == group_perm.operation
+
+
+# returns tuples of (parent ID, topic ID, topic name, perm type)
+def all_permissions_for_user(user):
+    possible_permissions = []
+    for membership in user.groupmembership_set.all():
+        group = membership.group
+        group_permissions = GroupKafkaPermission.objects.filter(principal=group).select_related('topic')
+        for permission in group_permissions:
+            if permission.operation==KafkaOperation.All:
+                for subpermission in KafkaOperation.__members__:
+                    possible_permissions.append((permission.id,permission.topic.id,permission.topic.name,subpermission))
+            else:
+                possible_permissions.append((permission.id,permission.topic.id,permission.topic.name,permission.operation))
+    # sort and eliminate duplicates
+    # sort on operation
+    possible_permissions.sort(key=lambda p: p[3])
+    # sort on topic names, because that looks nice for users, but since there is a bijection 
+    # between topic names and IDs this will place all matching topic IDs together in blocks 
+    # in some order
+    possible_permissions.sort(key=lambda p: p[2])
+
+    def equivalent(p1, p2):
+        return p1[1] == p2[1] and p1[3]==p2[3];
+    
+    # lack of an obvious analogue to std::unique makes this awkward, and not in-place
+    dedup = []
+    last = None
+    for p in possible_permissions:
+        if last is None or not equivalent(last,p):
+            dedup.append(p)
+            last=p
+    
+    return dedup
```

Comment:
This version does not work, unfortunately, because (as far as I can tell) `set` does not support user defined hash/comparison functions, and so it will never identify any duplicates under the definition we need (given by `equivalent`). Doing the deduplication after the sorting is then useful because it allows doing it in linear time, rather than quadratic, even with just a plain list.

Reply:
Ah, right - you only want to deduplicate on the 2nd and 4th fields of each tuple. Rats.

For what it's worth, `set` _does_ support user-defined hash functions - they should be the `__hash__` method of the provided objects. But making a custom type just so you can support `__hash__` just so you can support `set()` seems pretty roundabout.

The way you have deduplication working here is fine. I might do it differently, which I'll type out here just because its kind of interesting to see the differences (maybe?) but it's not blocking at all. I'd probably lean on Python's dictionary types to do the deduplication in the first pass, as you're assembling the set of permissions. I'd do it with a `collections.defaultdict` just because it makes things a smidge more tidy. Something like this:

```python
# returns tuples of (parent ID, topic ID, topic name, perm type)
def all_permissions_for_user(user):
    # Dictionary of topic -> operation -> permission granting that operation for that topic
    permissions = defaultdict(dict)

    memberships = user.groupmembership_set.all()
    for membership in memberships:
        group = membership.group
        group_permissions = GroupKafkaPermission.objects.filter(principal=group).select_related('topic')
        for permission in group_permissions:
            topic = permission.topic
            if permission.operation == KafkaOperation.All:
                for operation in KafkaOperation.__members__:
                    permissions[topic][operation] = permission
            else:
                permissions[topic][permission.operation] = permission

    # Sort by topic name
    sorted_permissions = sorted(permissions, key=lambda topic: topic.name)

    # Within each topic, sort by operation name
    for topic in sorted_permissions:
        sorted_permissions[topic] = sorted(sorted_permissions[topic], key=lambda operation: operation)

    # Reshape into list of tuples (Maybe should just return the dictionary structure?)
    result = []
    for topic in sorted_permissions:
        topic_perms = sorted_permissions[topic]
        for operation, permission in topic_perms.items():
            result.append((permission.id, topic.id, topic.name, operation))

    return result
```